internal/mode/static/nginx/config/servers.go

```diff
@@ -353,81 +396,110 @@ func initializeExternalLocations(
 		if !trailingSlashPrefixPathExists {
 			externalLocTrailing := http.Location{
 				Path: externalLocPath + "/",
+				Type: locType,
 			}
 			extLocations = append(extLocations, externalLocTrailing)
 		}
 		if !exactPathExists {
 			externalLocExact := http.Location{
 				Path: exactPath(externalLocPath),
+				Type: locType,
 			}
 			extLocations = append(extLocations, externalLocExact)
 		}
 	} else {
 		externalLoc := http.Location{
 			Path: externalLocPath,
+			Type: locType,
 		}
 		extLocations = []http.Location{externalLoc}
 	}
 
 	return extLocations
 }
 
+func getLocationTypeForPathRule(rule dataplane.PathRule) http.LocationType {
+	if needsInternalLocations(rule) {
+		return http.RedirectLocationType
+	}
+
+	return http.ExternalLocationType
+}
+
 func initializeInternalLocation(
 	pathruleIdx,
 	matchRuleIdx int,
 	match dataplane.Match,
+	grpc bool,
 ) (http.Location, routeMatch) {
-	path := fmt.Sprintf("@rule%d-route%d", pathruleIdx, matchRuleIdx)
-	return createMatchLocation(path), createRouteMatch(match, path)
+	path := fmt.Sprintf("%s-rule%d-route%d", http.InternalRoutePathPrefix, pathruleIdx, matchRuleIdx)
+	return createMatchLocation(path, grpc), createRouteMatch(match, path)
 }
 
-// updateLocationsForFilters updates the existing locations with any relevant filters.
-func updateLocationsForFilters(
+func updateLocationForFilters(
```

Comment:
I think this function needs to be renamed. It's now doing a lot more than adding filters...